brandon_tan/partner_selector.py

```diff
@@ -0,0 +1,483 @@
+            # and hence append it back here.
+            # Append (instead of insert) was chosen due to O(1) complexity.
+            partners.append(self.target_ticker)
+
+            partners_ranked_values = self.ranked_returns_data[partners].values
+
+            # Calling the private function _diagonal_distance to compute
+            # distance between point and the hyper diagonal.
+            total_euclidean_distance = np.sum(np.apply_along_axis(self._diagonal_distance, 1, partners_ranked_values))
+
+            if total_euclidean_distance < lowest_euclidean_distance:
+                lowest_euclidean_distance = total_euclidean_distance
+                best_Q = partners
+
+            partner_data[i] = {
+                'Grouping': partners,
+                'Total Euclidean Distance': total_euclidean_distance
+            }
+
+        self.partner_data = partner_data
+        self.best_Q = best_Q
+
+        return best_Q
+
+    def _diagonal_distance(self, vector: np.ndarray):
+        """
+        Helper private function to calculate the euclidean distance between a given point
+        and the hyper diagonal of the space.
+
+        The distance is computed by taking a perpendicular point on the hyper diagonal
+        as v1 = (k,k,k,k). Our point of interest is v2 = (w,y,x,z). Solving for the first derivative
+        of the distance d(v1,v2), we obtain a formula to compute the distance.
+
+        :param vector: The coordinates of the point.
+        :type vector: np.ndarray
+        :return: Euclidean distance between point and hyper diagonal of space.
+        :rtype: np.float
+        """
+        # Using a loop so that if the user chooses higher dimensions,
+        # this will still work, rather than a hardcoded formula.
+        euclidean_distance = 0
+        for i in vector:
+            # Math can be found in docstring.
+            mask = vector != i
+            other_terms_sum = sum(vector[mask])
+            euclidean_distance += ((3 * i - other_terms_sum) / 4) ** 2
+        return euclidean_distance
+
+    def print_info(self) -> pd.DataFrame:
+        """
+        Prints descriptive statistics of the selection process.
+        This is inspired by pd.describe().
+
+        :raises Exception: If selection process has not been done, an error will be raised.
+        :return: Dataframe containing descriptive statistics of the selection process.
+        :rtype: pd.DataFrame
+        """
+        if self.partner_data is None:
+            raise Exception("Potential parters have not been identified yet. Please run select() before this method.")
+
+        info_df = pd.DataFrame.from_dict(self.partner_data).transpose()
+        info_df = info_df[info_df.columns[1]]
+        info_df = info_df.astype(float)
+        info_df = pd.DataFrame(info_df.describe())
+        return info_df
+
+    def plot_selected_partners(self):
+        """
+        Plots the historical log prices of the best partners that was identified
```

Comment:
In the write up you mention that you are not sure what some good visualizations would be. It would be good to show visually, how the various techniques extract different properties from the data. Something to highlight the differences, which technique is better, something along those lines.